programs/bpf_loader/src/lib.rs

```diff
@@ -182,73 +282,336 @@ fn process_instruction_general(
+            }
+            if &program.owner()? != program_id {
+                log!(logger, "Program account not owned by loader");
+                return Err(InstructionError::IncorrectProgramId);
+            }
+            if let UpgradeableLoaderState::Program {
+                programdata_address,
+            } = program.state()?
+            {
+                if programdata_address != *programdata.unsigned_key() {
+                    log!(logger, "Program and ProgramData account mismatch");
+                    return Err(InstructionError::InvalidArgument);
+                }
+            } else {
+                log!(logger, "Invalid Program account");
+                return Err(InstructionError::InvalidArgument);
+            }
+
+            // Verify Buffer account
+
+            if UpgradeableLoaderState::Buffer != buffer.state()? {
+                log!(logger, "Invalid Buffer account");
+                return Err(InstructionError::InvalidInstructionData);
+            }
+
+            let buffer_data_offset = UpgradeableLoaderState::buffer_data_offset()?;
+            let buffer_data_len = buffer.data_len()?.saturating_sub(buffer_data_offset);
+            let programdata_data_offset = UpgradeableLoaderState::programdata_data_offset()?;
+            let programdata_balance_required = 1.max(
+                rent.minimum_balance(UpgradeableLoaderState::programdata_len(buffer_data_len)?),
+            );
+
+            // Verify ProgramData account
+
+            if programdata.data_len()? < UpgradeableLoaderState::programdata_len(buffer_data_len)? {
+                log!(logger, "ProgramData account not large enough");
+                return Err(InstructionError::AccountDataTooSmall);
+            }
+            if programdata.lamports()? + buffer.lamports()? < programdata_balance_required {
+                log!(logger, "Buffer account balance too low to fund upgrade");
+                return Err(InstructionError::InsufficientFunds);
+            }
+            if let UpgradeableLoaderState::ProgramData {
+                slot: _,
+                upgrade_authority_address,
+            } = programdata.state()?
+            {
+                if upgrade_authority_address != Some(*authority.unsigned_key()) {
+                    log!(logger, "Upgrade authority not present");
                     return Err(InstructionError::MissingRequiredSignature);
                 }
-                let offset = offset as usize;
-                let len = bytes.len();
-                if program.data_len()? < offset + len {
-                    log!(
-                        logger,
-                        "Write overflow: {} < {}",
-                        program.data_len()?,
-                        offset + len
-                    );
-                    return Err(InstructionError::AccountDataTooSmall);
+                if authority.signer_key().is_none() {
+                    log!(logger, "Upgrade authority did not sign");
```

Comment:
I like all the logging for the error conditions.  This is a good precedent for the native programs given how painful it currently is to decode custom error codes

Reply:
What do you think about providing even more information, like for example reporting the wrong balance or size values using `format!`?